pkg/sql/opt/optbuilder/scope.go

```diff
@@ -478,7 +478,7 @@ func (s *scope) resolveType(expr tree.Expr, desired *types.T) tree.TypedExpr {
 // desired type.
 func (s *scope) resolveAndRequireType(expr tree.Expr, desired *types.T) tree.TypedExpr {
 	expr = s.walkExprTree(expr)
-	texpr, err := tree.TypeCheckAndRequire(s.builder.ctx, expr, s.builder.semaCtx, desired, s.context.String())
+	texpr, err := tree.TypeCheckAndRequire(s.builder.ctx, expr, s.builder.semaCtx, desired, s.context.String(), true)
```

Comment:
I don't think it should be necessary to change `tree.TypeCheckAndRequire` now that https://github.com/cockroachdb/cockroach/pull/81863 is merged. Can you try passing `false` instead of `true` here and check if any tests break? If the tests pass, you can revert all changes related to `tree.TypeCheckAndRequire`. If they don't, we can do some more sleuthing to figure out the problem.